troubadix/helper/linguistic_exception_handler.py

```diff
@@ -0,0 +1,162 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import re
+from abc import ABC, abstractmethod
+from typing import List, Tuple, Union
+
+
+class LinguisticCheck(ABC):
+    @abstractmethod
+    def execute(self, file, content):
+        pass
+
+
+class FileCheck(LinguisticCheck):
+    def __init__(self, file: str) -> None:
+        self.file = file
+
+    def execute(self, file: str, content: str) -> bool:
+        return self.file in file
+
+
+class FilesCheck(LinguisticCheck):
+    def __init__(self, files: List[str]) -> None:
+        self.files = files
+
+    def execute(self, file: str, content: str):
+        return any([_file in file for _file in self.files])
+
+
+class FilePatternCheck(LinguisticCheck):
+    def __init__(self, file_pattern, flags=0) -> None:
+        self.file_pattern = re.compile(file_pattern, flags=flags)
+
+    def execute(self, file: str, content: str):
+        return bool(self.file_pattern.search(file))
+
+
+class TextCheck(LinguisticCheck):
+    def __init__(self, text: str) -> None:
+        self.text = text
+
+    def execute(self, file: str, content: str):
+        return self.text in content
+
+
+class PatternCheck(LinguisticCheck):
+    def __init__(self, pattern, flags=0) -> None:
+        self.pattern = re.compile(pattern, flags=flags)
+
+    def execute(self, file: str, content: str):
+        return bool(self.pattern.search(content))
+
+
+class PatternsCheck(LinguisticCheck):
+    def __init__(
+        self,
+        patterns: Union[List[str], List[Tuple[str, re.RegexFlag]]],
+    ) -> None:
+        # Originally I tried isinstance(patterns, Tuple[...]) for testing,
+        # but that throws an exception because Tuple[...]
+        # is only valid for type hinting :/
+        if isinstance(patterns[0], Tuple):
+            self.patterns = [
+                re.compile(pattern, flags=flags) for pattern, flags in patterns
+            ]
+        else:
+            self.patterns = [re.compile(pattern) for pattern in patterns]
+
+    def execute(self, file, content):
+        return any(bool(pattern.search(content)) for pattern in self.patterns)
+
+
+class CompositeCheck(LinguisticCheck):
+    def __init__(
+        self, file_check: LinguisticCheck, text_check: LinguisticCheck
+    ) -> None:
+        self.file_check = file_check
+        self.text_check = text_check
+
+    def execute(self, file: str, content: str):
+        return self.file_check.execute(
+            file, content
+        ) and self.text_check.execute(file, content)
```

Comment:
```suggestion
    def __init__(
        self, *checks: Iterable[LinguisticCheck]
    ) -> None:
        self.checks = checks

    def execute(self, file: str, content: str) -> bool:
        for check in self.checks:
            if not check.execute(file, content):
                return False
        return True
```

Reply:
I think we could go even easiert with (written by hand in here right now, I think the gist is clear):

```
def execute(self, file: str, content: str) -> bool:
  return all(check.execute(file, content) for check in checks)
```

?

Reply:
Yes indeed as a generator this is nice. Otherwise you would iterate twice over the list.

Reply:
Implemented in latest push